tasks/GroundFollower.cpp

```diff
@@ -0,0 +1,133 @@
+/* Generated from orogen/lib/orogen/templates/tasks/Task.cpp */
+
+#include "GroundFollower.hpp"
+//#include <math.h>
+
+using namespace auv_control;
+
+GroundFollower::GroundFollower(std::string const& name, TaskCore::TaskState initial_state)
+    : GroundFollowerBase(name, initial_state)
+{
+}
+
+GroundFollower::GroundFollower(std::string const& name, RTT::ExecutionEngine* engine, TaskCore::TaskState initial_state)
+    : GroundFollowerBase(name, engine, initial_state)
+{
+}
+
+GroundFollower::~GroundFollower()
+{
+}
+
+
+
+/// The following lines are template definitions for the various state machine
+// hooks defined by Orocos::RTT. See GroundFollower.hpp for more detailed
+// documentation about them.
+
+bool GroundFollower::configureHook()
+{
+    if (! GroundFollowerBase::configureHook())
+        return false;
+
+    new_altimeter_timeout = base::Timeout(_altimeter_timeout.get());
+    new_depth_timeout = base::Timeout(_depth_timeout.get());
+    altimeter_dropout_timeout = base::Timeout(_altimeter_dropout_timeout.get());
+
+    distance_to_ground_cmd = _distance_to_ground.get();
+    if(distance_to_ground_cmd <= 0){
+        exception(INVALID_TARGET_DEPTH_CONFIG);
+        return false;
+    }
+    return true;
+}
+bool GroundFollower::startHook()
+{
+    if (! GroundFollowerBase::startHook())
+        return false;
+
+    //last_valid_ground_position = NAN;
+    new_altimeter_timeout.restart();
+    new_depth_timeout.restart();
+    altimeter_dropout_timeout.restart();
+    return true;
+}
+void GroundFollower::updateHook()
+{
+    GroundFollowerBase::updateHook();
+
+    RTT::FlowStatus altimeter_status = _altimeter.readNewest(altimeter);
+    RTT::FlowStatus depth_status = _depth.readNewest(depth);
+
+    if(altimeter_status == RTT::NoData){
```

Comment:
The timeouts currently only would check if there was never a samples received. In case you want to check also if there have been no new samples, you have to check RTT::OldData as well or only check for RTT::NewData.

Reply:
updated the states transitions, now when no samples are received at all the task will wait for the timeout with a runtime state (NO_'SENSOR'_READING) then go into exception ('SENSOR'_TIMEOUT). If the data is old the task will stay running and write out a command using the old data for the timeout specified then go into exception.